File: src/my_class_pkg/scripts/ros_action_server.py
#!/usr/bin/env python3
# 修正：import 后加空格，规范导入语句
import rospy
import actionlib
# 修正：单行完成导入，避免换行错误
from my_class_pkg.msg import MyActionAction, MyActionGoal, MyActionResult, MyActionFeedback

class MyActionServer(object):
    # 修正：函数名和括号间加空格，规范缩进
    def __init__(self, name):
        self._action_name = name
        # 初始化 Action 服务器（auto_start=False 手动启动）
        self._as = actionlib.SimpleActionServer(
            self._action_name, 
            MyActionAction,
            execute_cb=self.executeCB,
            auto_start=False
        )
        self._as.start()  # 启动 Action 服务器
        # 初始化反馈和结果消息
        self._feedback = MyActionFeedback()
        self._result = MyActionResult()

    # 核心回调函数：处理客户端的目标请求
    def executeCB(self, goal):
        r = rospy.Rate(1)  # 1Hz 循环频率
        success = True

        # 模拟耗时任务（循环10次，进度每次+10%）
        for i in range(1, 11):
            # 检查是否被客户端取消
            if self._as.is_preempt_requested():
                rospy.loginfo('{}: Preempted'.format(self._action_name))
                self._as.set_preempted()  # 标记任务被取消
                success = False
                break

            # 更新进度反馈并发布
            self._feedback.progress = i * 10
            rospy.loginfo(
                '{}: Executing, progress = {}%'.format(
                    self._action_name, 
                    self._feedback.progress
                )
            )
            self._as.publish_feedback(self._feedback)
            r.sleep()  # 按频率休眠

        # 任务完成，发送最终结果
        if success:
            self._result.success = True
            rospy.loginfo('{}: Succeeded'.format(self._action_name))
            self._as.set_succeeded(self._result)  # 标记任务成功

if __name__ == '__main__':
    # 初始化 ROS 节点
    rospy.init_node('my_action_server_py')
    # 创建 Action 服务器实例（服务名：my_action）
    server = MyActionServer('my_action')
    # 保持节点运行
    rospy.spin()

File: src/my_class_pkg/src/ros_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <my_class_pkg/MyActionAction.h>

class MyActionServer {
protected:
    ros::NodeHandle nh_;
    // 修正：as_ 变量名和类型之间加空格（规范写法）
    actionlib::SimpleActionServer<my_class_pkg::MyActionAction> as_;
    // 修正1：std::string 和 action_name_ 之间加空格
    std::string action_name_;
    my_class_pkg::MyActionFeedback feedback_;
    my_class_pkg::MyActionResult result_;

public:
    // 构造函数：初始化 Action 服务器
    MyActionServer(std::string name) :
        as_(nh_, name, boost::bind(&MyActionServer::executeCB, this, _1), false),
        action_name_(name) 
    {
        as_.start(); // 启动 Action 服务器
    }

    ~MyActionServer(void) {}

    // 核心回调函数：处理客户端的目标请求
    void executeCB(const my_class_pkg::MyActionGoalConstPtr& goal) {
        // 修正2：拼写错误 Rater→Rate，且声明变量名 r
        ros::Rate r(1); // 1Hz 循环频率
        bool success = true;

        // 模拟耗时任务（循环10次，每次进度+10%）
        for (int i = 1; i <= 10; i++) {
            // 检查是否被客户端取消
            if (as_.isPreemptRequested() || !ros::ok()) {
                ROS_INFO("%s: Preempted", action_name_.c_str());
                as_.setPreempted(); // 标记任务被取消
                success = false;
                break;
            }

            // 更新反馈（进度值）并发布
            feedback_.progress = i * 10;
            ROS_INFO("%s: Executing, progress = %.1f%%", action_name_.c_str(), feedback_.progress);
            as_.publishFeedback(feedback_);

            // 修正3：变量 r 已声明，可正常调用 sleep()
            r.sleep();
        }

        // 任务完成，发送最终结果
        if (success) {
            result_.success = true;
            ROS_INFO("%s: Succeeded", action_name_.c_str());
            as_.setSucceeded(result_); // 标记任务成功
        }
    }
};

int main(int argc, char** argv) {
    // 初始化 ROS 节点
    ros::init(argc, argv, "my_action_server");
    // 创建 Action 服务器实例（服务名：my_action）
    MyActionServer server("my_action");
    // 循环处理回调
    ros::spin();

    return 0;
}
